README.md
# avoriaz [![Build Status](https://travis-ci.org/eddyerburgh/avoriaz.svg?branch=master)](https://travis-ci.org/eddyerburgh/avoriaz)


> a Vue.js testing utility library


## Installation

```
npm install --save-dev avoriaz
```

## Documentation

[Visit the docs](https://eddyerburgh.gitbooks.io/avoriaz/content/)

## Examples

- [Example using karma and mocha](https://github.com/eddyerburgh/avoriaz-karma-mocha-example)
- [Example using karma and jasmine](https://github.com/eddyerburgh/avoriaz-karma-jasmine-example)
- [Example using Jest](https://github.com/eddyerburgh/avoriaz-jest-example)
- [Example using mocha-webpack](https://github.com/eddyerburgh/avoriaz-mocha-example)
- [Example using tape](https://github.com/eddyerburgh/avoriaz-tape-example)
- [Example using ava](https://github.com/eddyerburgh/avoriaz-ava-example)

##### Assert wrapper contains a child
```js
import { mount } from 'avoriaz'
import Foo from './Foo.vue'

const wrapper = mount(Foo)
expect(wrapper.contains('.bar')).to.equal(true)
```

##### Shallow render components
```js
import { shallow } from 'avoriaz'
import Foo from './Foo.vue'
import Bar from './Bar.vue'

const wrapper = shallow(Foo)
expect(wrapper.contains(Bar)).to.equal(true)
```

##### Assert style is rendered
```js
const button = wrapper.find('div > button .button-child')[0]
expect(button.hasStyle('color', 'red')).to.equal(true)
```

##### Assert method is called when DOM event is triggered
```js
const clickHandler = sinon.stub()
const wrapper = mount(Foo, {
  propsData: { clickHandler }
})
wrapper.find('div .bar')[0].trigger('click')
expect(clickHandler.called).to.equal(true)
```

##### Assert wrapper contains text
```js
const title = wrapper.find('h1.title')[0]
expect(title.text()).to.equal('some text')
```

##### Inject globals
```js
const $route = { path: 'http://www.example-path.com' }
const wrapper = mount(Foo, { 
    globals: {
        $route
    }
})
expect(wrapper.vm.$route.path).to.equal($route.path)
```

##### Inject slots
```js
const wrapper = mount(Foo, { 
    slots: {
        default: Foo
    }
})
```

##### Set data
```js
wrapper.setData({
  someData: 'some data'
})

expect(wrapper.vm.someData).to.equal('some data')
```

##### Update props
```js
wrapper.setProps({
  someProp: 'some prop',
  anotherProp: 'another prop'
})
```

For more examples, [see the docs](https://eddyerburgh.gitbooks.io/avoriaz/content/)


CHANGELOG.md
4.2.0 / 09-17-2017
=================
* [New] Add renderDefaultSlot option to shallow

4.1.0 / 09-04-2017
=================
* [New] Add listeners option

4.0.0 / 08-30-2017
=================
* [New] Add getAttribute method
* [New] Add getProp method

3.6.0 / 08-29-2017
=================
  * [New] Add provide option to mount

3.5.0 / 08-24-2017
=================
  * [New] Add .amd and .umd builds

3.4.0 / 08-22-2017
=================
  * [New] Add setMethods method

3.3.1 / 08-20-2017
=================
  * [Fix] Add correct return from find in type definitions

3.3.0 / 08-20-2017
=================
  * [New] Add value method

3.2.1 / 08-20-2017
=================
  * [New] Add attrs object by default

3.2.0 / 08-04-2017
=================
  * [New] Add option to pass Wrappers to slots

3.1.0 / 08-02-2017
=================
  * [New] Add Types support

3.0.0 / 07-29-2017
=================
  * [Fix] Remove update from trigger method
  * Remove style method
  * Remove use method
  * Remove dispatch method
  * Remove simulate method

2.6.5 / 07-24-2017
=================
  * [Build] Revert to previous build

2.6.3 / 07-22-2017
=================
  * [Fix] Add polyfill for Event in IE10, 11

2.6.2 / 07-22-2017
=================
  * [Fix] Fixing broken build system

2.6.0 / 07-22-2017
=================
  * [Build] Use rollup for dist file

2.5.0 / 07-18-2017
=================
  * [New] Add option to pass `attrs` to vm

2.4.4 / 07-16-2017
=================
  * [Fix] Update all props before running watchers in setProps

2.4.3 / 07-11-2017
=================
  * [Fix] Stub global components in shallow

2.4.2 / 07-03-2017
=================
  * [Fix] Only keep specific properties from stubbed components in shallow

2.4.1 / 06-30-2017
=================
  * [Fix] Only run watcher for relevant expression in `setProps`

2.4.0 / 06-26-2017
=================
  * [New] Add `mount.options.instance` to pass scoped instance

2.3.0 / 06-23-2017
=================
  * [New] Add `mount.options.context` to handle functional components

2.2.1 / 06-23-2017
=================
  * [Fix] Clone components before stubbing in `shallow`
  * [Fix] Update before calling trigger on element

2.2.0 / 06-22-2017
=================
  * [New] Add first method

2.1.3 / 06-21-2017
=================
  * [Fix] Run watchers in setProps

2.1.2 / 06-21-2017
=================
  * [Fix] Don't stub root vm lifecycle hooks

2.1.1 / 06-20-2017
=================
  * [Fix] Stub child lifecycle hooks in shallow method

2.1.0 / 06-19-2017
=================
  * [New] Add shallow method

2.0.0 / 06-12-2017
=================
  * [Refactor] Remove emit behavior from trigger. Breaking change

1.16.0 / 06-12-2017
=================
  * [Fix] Emit event on child vue instance
  * [Refactor] Use matches in place of sizzle
  * [New] Deprecate dispatch method
  * [New] Add trigger method

1.15.0 / 05-31-2017
=================
  * [New] Use sizzle to match selectors in find

1.14.0 / 05-26-2017
=================
  * [New] Add dispatch method
  * [New] Deprecate simulate method
  * [Fix] List Vue as a peer dependency

1.13.3 / 05-14-2017
=================
  * [Fix] Improve error handling of unnamed components used in .find()

1.13.2 / 05-13-2017
=================
  * [Docs] Add improved examples to README

1.13.1 / 05-08-2017
=================
  * [Fix] Delete cached Ctor before mount

1.13.0 / 05-08-2017
=================
  * [New] Add option to pass Components to slots
  * [New] Add option to pass globals in mount method
  * [New] Add setProps wrapper method

1.12.1 / 05-07-2017
=================
  * [New] Remove development mode warning log
  * [Fix] Handle slots as arrays

1.12.0 / 05-05-2017
=================
  * [New] Add slots mount option

1.11.2 / 04-28-2017
=================
  * [Fix] Add undocumented options argument to use method

1.11.1 / 04-14-2017
=================
  * [Fix] Mount entire component in hasStyle to accurately detect CSS

1.11.0 / 04-14-2017
=================
  * [New] Add attachToDocument option
  * [New] Use getters to retrieve vNode
  * [Fix] Remove duplicate vNodes by comparing element node
  * [Docs] Standardize docs
  * [Docs] Add guide to using with vuex

1.10.0 / 04-09-2017
=================
  * [New] Support key modifiers in simulate method
  * [Docs] Add examples to docs

1.9.4 / 03-30-2017
=================
  * [Fix] Change propsData warning message to give a correct example

1.9.3 / 03-29-2017
=================
  * [Docs] Add examples to docs
  * [Docs] Fix typo in README

1.9.2 / 03-27-2017
=================
  * [Docs] Add examples to docs
  * [Feature] Log warning when computed, propsData and methods are called with details on unbound this

1.9.1 / 03-16-2017
=================
  * [Bug] Update vNode when setData is called

1.9.0 / 03-15-2017
=================
  * Add instance method

1.8.3 / 03-05-2017
=================
  * Expose use method

1.8.2 / 03-05-2017
=================
  * Update Vue and vue-compiler version

1.8.1 / 03-05-2017
=================
  * Add Vue to list of dependencies

1.8.0 / 03-05-2017
=================
  * Move Vue to list of peer dependencies

1.7.2 / 03-05-2017
=================
  * Remove use method due to bug

1.7.1 / 03-04-2017
=================
  * Move Vue to list of dependencies

1.7.0 / 03-04-2017
=================
  * Add setData method
  * Add destroy method

1.6.0 / 02-12-2017
=================
  * Add use method to add plugins to avoriaz Vue instance

1.5.2 / 02-08-2017
=================
  * Throw error if avoriaz does not have access to document

1.5.1 / 02-08-2017
=================
  * Remove babel-runtime from project

1.5.0 / 02-03-2017
=================
  * Add .hasStyle method
  * Add .hasAttribute method
  * Deprecate .style method
  * Reinstate unit tests running in jsdom

1.4.0 / 02-01-2017
=================
  * Get computedStyles instead of element.style in style method

1.3.1 / 02-01-2017
=================
  * Fix npm publishing issue

1.3.0 / 02-01-2017
=================

src/Wrapper.js
// @flow

import findMatchingVNodes from './lib/find-matching-vnodes';
import {
  findVueComponents,
  vmCtorMatchesName,
} from './lib/vm';
import {
  isValidSelector,
} from './lib/validators';
import VueWrapper from './VueWrapper';
import {
  error,
  warn,
} from './lib/logger';

export default class Wrapper implements WrapperInterface {
  vNode: VNode;
  vm: Component;
  element: HTMLElement;
  update: Function;
  mountedToDom: boolean;
  isVueComponent: boolean;

  constructor(vNode: VNode, update: Function, mountedToDom: boolean) {
    this.vNode = vNode;
    this.element = vNode.elm;
    this.update = update;
    this.mountedToDom = mountedToDom;
  }

  /**
   * Returns instances computed object
   *
   * @returns {Object}
   */
  computed() {
    if (!this.isVueComponent) {
      error('wrapper.computed() can only be called on a Vue instance');
    }

    warn('functions returned by computed() will not have this bound to the vue instance. Calling a computed function that uses this will result in an error. You can access computed functions by using the vue instance. e.g. to call a computed function named compFunc, call wrapper.vm.compFunc. See https://github.com/eddyerburgh/avoriaz/issues/15');

    return this.vm.$options.computed;
  }

  /**
   * Checks if wrapper contains provided selector.
   *
   * @param {String} selector
   * @returns {Boolean}
   */
  contains(selector: Selector) {
    if (!isValidSelector(selector)) {
      error('wrapper.contains() must be passed a valid CSS selector or a Vue constructor');
    }

    if (typeof selector === 'object') {
      const vm = this.vm || this.vNode.context.$root;
      return findVueComponents(vm, selector.name).length > 0;
    }

    return this.element.querySelectorAll(selector).length > 0;
  }

  /**
   * Returns instances data object
   *
   * @returns {Object}
   */
  data() {
    if (!this.isVueComponent) {
      error('wrapper.data() can only be called on a Vue instance');
    }

    return this.vm._data;
  }

  /**
   * Calls destroy on vm
   */
  destroy() {
    if (!this.isVueComponent) {
      error('wrapper.destroy() can only be called on a Vue instance');
    }

    if (this.vm.$el.parentNode) {
      this.vm.$el.parentNode.removeChild(this.vm.$el);
    }

    this.vm.$destroy();
  }

  /**
   * Finds every node in the mount tree of the current wrapper that matches the provided selector.
   *
   * @param {String|Object} selector
   * @returns {VueWrapper||VueWrapper[]}
   */
  find(selector: Selector) {
    if (!isValidSelector(selector)) {
      error('wrapper.find() must be passed a valid CSS selector or a Vue constructor');
    }

    if (typeof selector === 'object') {
      if (!selector.name) {
        error('.find() requires component to have a name property');
      }
      const vm = this.vm || this.vNode.context.$root;
      const components = findVueComponents(vm, selector.name);
      return components.map(component => new VueWrapper(component, this.mountedToDom));
    }

    const nodes = findMatchingVNodes(this.vNode, selector);

    return nodes.map(node => new Wrapper(node, this.update, this.mountedToDom));
  }

  /**
   * Returns the first node that matches the provided selector.
   *
   * @param {String|Object} selector
   * @returns {VueWrapper}
   */
  first(selector: Selector) {
    const nodes = this.find(selector);

    if (!nodes.length) {
      error('wrapper.first() has no matches with the given selector');
    }

    return nodes[0];
  }

  /**
   * Returns value of specified attribute of wrapper
   *
   * @param {String} attribute - attribute to assert
   * @returns {String}
   */
  getAttribute(attribute: string) {
    if (typeof attribute !== 'string') {
      error('wrapper.getAttribute() must be passed a string');
    }

    if (!this.hasAttribute(attribute)) {
      error(`wrapper has no attribute called ${attribute}`);
    }

    return this.element.getAttribute(attribute);
  }

  /**
   * Returns prop value of a Vue instance
   *
   * @param {String} propName - prop name to assert
   * @returns {*}
   */
  getProp(propName: string) {
    if (!this.isVueComponent) {
      error('wrapper.getProp() can only be called on a Vue instance');
    }

    if (typeof propName !== 'string') {
      error('wrapper.getProp() must be passed a string');
    }

    const propValue = this.vm.$props[propName];
    if (typeof propValue === 'function') {
      warn('functions returned by getProp() will not have this bound to the vue instance. Calling a propsData function that uses this will result in an error. You can access propsData functions by using the vue instance. e.g. to call a method function named propsDataFunc, call wrapper.vm.$props.propsDataFunc(). See https://github.com/eddyerburgh/avoriaz/issues/15');
    }
    return propValue;
  }

  /**
   * Asserts wrapper has an attribute
   *
   * @param {String} attribute - attribute to assert
   * @returns {Boolean}
   */
  hasAttribute(attribute: string, value?: string) {
    if (typeof attribute !== 'string') {
      error('wrapper.hasAttribute() must be passed attribute as a string');
    }

    if (arguments.length === 1) {
      return this.element.hasAttribute(attribute);
    }

    if (typeof value !== 'string') {
      error('wrapper.hasAttribute() must be passed value as a string');
    }

    warn('wrapper.hasAttribute(attribute, value) is deprecated in place of a new syntax and the old syntax will be removed from future versions. Instead, we encourage you to use getAttribute(attribute) === value to assert attribute value for better error report. For detailed information, see: https://github.com/eddyerburgh/avoriaz/issues/100');

    return this.element.getAttribute(attribute) === value;
  }

  /**
   * Asserts wrapper has a class name
   *
   * @param {String} className - class name to assert
   * @returns {Boolean}
   */
  hasClass(className: string) {
    if (typeof className !== 'string') {
      error('wrapper.hasClass() must be passed a string');
    }

    return this.element.className.split(' ').indexOf(className) !== -1;
  }

  /**
   * Checks if wrapper has a style with value
   *
   * @param {String} style - style to assert
   * @param {String} value - value to assert style contains
   * @returns {Boolean}
   */
  hasStyle(style: string, value: string) {
    if (typeof style !== 'string') {
      error('wrapper.hasStyle() must be passed style as a string');
    }

    if (typeof value !== 'string') {
      error('wrapper.hasClass() must be passed value as string');
    }

    /* istanbul ignore next */
    if (navigator.userAgent.includes && navigator.userAgent.includes('node.js')) {
      warn('wrapper.hasStyle is not fully supported when running jsdom - only inline styles are supported');
    }
    const body = document.querySelector('body');
    const mockElement = document.createElement('div');
    // $FlowIgnore
    const mockNode = body.insertBefore(mockElement, null);
    // $FlowIgnore
    mockElement.style[style] = value;

    if (!this.mountedToDom) {
      const vm = this.vm || this.vNode.context.$root;
      // $FlowIgnore
      body.insertBefore(vm.$root._vnode.elm, null);
    }

    const elStyle = window.getComputedStyle(this.element)[style];
    const mockNodeStyle = window.getComputedStyle(mockNode)[style];
    return elStyle === mockNodeStyle;
  }

  /**
   * Returns HTML of element as a string
   *
   * @returns {String} HTML of wrapper element
   */
  html() {
    const tmp = document.createElement('div');
    tmp.appendChild(this.element);
    return tmp.innerHTML;
  }

  /**
   * Returns Vue instance
   *
   * @returns {String} HTML of wrapper element
   */
  instance() {
    if (!this.isVueComponent) {
      error('wrapper.instance() can only be called on a Vue instance');
    }

    return this.vm;
  }

  /**
   * Checks if node matches selector
   *
   * @param {String} selector - selector to check node is
   * @returns {Boolean}
   */
  is(selector: Selector) {
    if (!isValidSelector(selector)) {
      error('wrapper.is() must be passed a valid CSS selector or a Vue constructor');
    }

    if (typeof selector === 'object') {
      if (!this.isVueComponent) {
        return false;
      }

      return vmCtorMatchesName(this.vm, selector.name);
    }

    if (selector[0] === '.') {
      return this.element.className.split(' ').indexOf(selector.substr(1)) !== -1;
    }

    if (selector[0] === '#') {
      return this.element.getAttribute('id') === selector.substr(1);
    }

    return this.element.tagName === selector.toUpperCase();
  }

  /**
   * Checks if node is empty
   *
   * @returns {Boolean}
   */
  isEmpty() {
    return this.vNode.children === undefined || this.vNode.children.length === 0;
  }

  /**
   * Returns instances methods object
   *
   * @returns {Object}
   */
  methods() {
    if (!this.isVueComponent) {
      error('wrapper.methods() can only be called on a Vue instance');
    }

    warn('functions returned by methods() will not have this bound to the vue instance. Calling a method that uses this will result in an error. You can access methods by using the vue instance. e.g. to call a method function named aMethod, call wrapper.vm.aMethod(). See https://github.com/eddyerburgh/avoriaz/issues/15');

    return this.vm.$options.methods;
  }

  /**
   * Returns name of component, or tag name if node is not a Vue component
   *
   * @returns {String}
   */
  name() {
    if (this.isVueComponent) {
      return this.vm.$options.name;
    }

    return this.vNode.tag;
  }

  /**
   * Sets vm methods
   *
   * @param {Object} methods - methods to set
   */
  setMethods(methods: Object) {
    if (!this.isVueComponent) {
      error('wrapper.setMethods() can only be called on a Vue instance');
    }

    Object.keys(methods).forEach((key) => {
      this.vm[key] = methods[key];
      this.vm.$options.methods[key] = methods[key];
    });
    this.update();
    this.vNode = this.vm._vnode;
  }

  /**
   * Returns instances propsData object
   *
   * @returns {String}
   */
  propsData() {
    if (!this.isVueComponent) {
      error('wrapper.propsData() can only be called on a Vue instance');
    }

    warn('functions returned by propsData() will not have this bound to the vue instance. Calling a propsData function that uses this will result in an error. You can access propsData functions by using the vue instance. e.g. to call a method function named propsDataFunc, call wrapper.vm.$props.propsDataFunc(). See https://github.com/eddyerburgh/avoriaz/issues/15');

    return this.vm.$props;
  }

  /**
   * Sets vm data
   *
   * @param {Object} data - data to set
   */
  setData(data: Object) {
    if (!this.isVueComponent) {
      error('wrapper.setData() can only be called on a Vue instance');
    }

    Object.keys(data).forEach((key) => {
      this.vm.$set(this.vm, [key], data[key]);
    });
    this.update();
    this.vNode = this.vm._vnode;
  }

  /**
   * Sets vm props
   *
   * @param {Object} props - props to set
   */
  setProps(props: Object) {
    if (!this.isVueComponent) {
      error('wrapper.setProps() can only be called on a Vue instance');
    }
    const vm = this.vm || this.vNode.context.$root;

    Object.keys(props).forEach((key) => {
      this.vm._props[key] = props[key];
    });
    Object.keys(props).forEach((key) => {
      vm._watchers.forEach((watcher) => {
        if (watcher.expression === key) { watcher.run(); }
      });
    });
    this.update();

    this.vNode = this.vm._vnode;
  }

  /**
   * Return text of wrapper element
   *
   * @returns {Boolean}
   */
  text() {
    return this.element.textContent;
  }

  /**
  * Return value of wrapper element
  *
  * @returns {String | Boolean}
  */
  value() {
    if (!this.element.value) return false;
    return this.element.value;
  }

  /**
     * Triggers a DOM event on wrapper
     *
     * @param {String} type - type of event
     * @returns {Boolean}
     */
  trigger(type: string) {
    if (typeof type !== 'string') {
      error('wrapper.trigger() must be passed a string');
    }

    const modifiers = {
      enter: 13,
      tab: 9,
      delete: 46,
      esc: 27,
      space: 32,
      up: 38,
      down: 40,
      left: 37,
      right: 39,
    };

    const event = type.split('.');

    let eventObject;

    // Fallback for IE10,11 - https://stackoverflow.com/questions/26596123
    if (typeof (Event) === 'function') {
      eventObject = new window.Event(event[0]);
    } else {
      eventObject = document.createEvent('Event');
      eventObject.initEvent(event[0], true, true);
    }

    if (event.length === 2) {
      // $FlowIgnore
      eventObject.keyCode = modifiers[event[1]];
    }

    this.element.dispatchEvent(eventObject);
    this.update();
  }
}
